expr/expr.go

```diff
@@ -3266,6 +3267,74 @@ func EvalExpr(e *expr, from, until int32, values map[MetricRequest][]*MetricData
 		}
 
 		return results, nil
+
+
+	case "cactiStyle": // cactiStyle(seriesList, system=None, units=None)
+		// Get the series data
+		original, err := getSeriesArg(e.args[0], from, until, values)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get the arguments
+		var system string = ""
+		var unit string = ""
+		if len(e.args) >= 2 {
+			system = e.args[1].valStr
+		}
+		if len(e.args) >= 3 {
+			unit = e.args[2].valStr
+		}
+
+		// Deal with each of the series
+		for _, a := range original {
+			var min string
+			var current string
+			var max string
+
+			// Calculate min, ignoring any absent values and inf (if we have a value)
+			// Using summarizeValues("min", ...) results in incorrect values, when absent
+			// values are present
```

Comment:
I think I've got this into 1 pass now, since the entire array needs to be walked to calculate min, calculating max and last should just work. This assumes the list is in time order for "last", but so does `summarizeValues("last",...)` due to slicing the last element.